test/e2e/framework/resource/resources.go

```diff
@@ -0,0 +1,210 @@
+	scalesGetter scaleclient.ScalesGetter,
+	ns, name string,
+	size uint,
+	wait bool,
+	kind schema.GroupKind,
+	gvr schema.GroupVersionResource,
+) error {
+	ginkgo.By(fmt.Sprintf("Scaling %v %s in namespace %s to %d", kind, name, ns, size))
+	if err := testutils.ScaleResourceWithRetries(scalesGetter, ns, name, size, gvr); err != nil {
+		return fmt.Errorf("error while scaling RC %s to %d replicas: %v", name, size, err)
+	}
+	if !wait {
+		return nil
+	}
+	return WaitForControlledPodsRunning(clientset, ns, name, kind)
+}
+
+// DeleteResourceAndWaitForGC deletes only given resource and waits for GC to delete the pods.
+func DeleteResourceAndWaitForGC(c clientset.Interface, kind schema.GroupKind, ns, name string) error {
+	ginkgo.By(fmt.Sprintf("deleting %v %s in namespace %s, will wait for the garbage collector to delete the pods", kind, name, ns))
+
+	rtObject, err := GetRuntimeObjectForKind(c, kind, ns, name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			framework.Logf("%v %s not found: %v", kind, name, err)
+			return nil
+		}
+		return err
+	}
+	selector, err := GetSelectorFromRuntimeObject(rtObject)
+	if err != nil {
+		return err
+	}
+	replicas, err := GetReplicasFromRuntimeObject(rtObject)
+	if err != nil {
+		return err
+	}
+
+	ps, err := testutils.NewPodStore(c, ns, selector, fields.Everything())
+	if err != nil {
+		return err
+	}
+
+	defer ps.Stop()
+	falseVar := false
+	deleteOption := metav1.DeleteOptions{OrphanDependents: &falseVar}
+	startTime := time.Now()
+	if err := testutils.DeleteResourceWithRetries(c, kind, ns, name, deleteOption); err != nil {
+		return err
+	}
+	deleteTime := time.Since(startTime)
+	framework.Logf("Deleting %v %s took: %v", kind, name, deleteTime)
+
+	var interval, timeout time.Duration
+	switch {
+	case replicas < 100:
+		interval = 100 * time.Millisecond
+	case replicas < 1000:
+		interval = 1 * time.Second
+	default:
+		interval = 10 * time.Second
+	}
+	if replicas < 5000 {
+		timeout = 10 * time.Minute
+	} else {
+		timeout = time.Duration(replicas/gcThroughput) * time.Second
+		// gcThroughput is pretty strict now, add a bit more to it
+		timeout = timeout + 3*time.Minute
+	}
+
+	err = waitForPodsInactive(ps, interval, timeout)
+	if err != nil {
+		return fmt.Errorf("error while waiting for pods to become inactive %s: %v", name, err)
+	}
+	terminatePodTime := time.Since(startTime) - deleteTime
+	framework.Logf("Terminating %v %s pods took: %v", kind, name, terminatePodTime)
+
+	// In gce, at any point, small percentage of nodes can disappear for
+	// ~10 minutes due to hostError. 20 minutes should be long enough to
+	// restart VM in that case and delete the pod.
+	err = waitForPodsGone(ps, interval, 20*time.Minute)
+	if err != nil {
+		return fmt.Errorf("error while waiting for pods gone %s: %v", name, err)
+	}
+	return nil
+}
+
+// waitForPodsGone waits until there are no pods left in the PodStore.
+func waitForPodsGone(ps *testutils.PodStore, interval, timeout time.Duration) error {
```

Comment:
Ima be a bit too verbose just for the sake of communication.
I was about to make a comment on `waitForPodsGone` and `waitForPodsInactive`.
The mechanics for these 2 functions is different than for `WaitForControlledPodsRunning` and `WaitForControlledPods`.
As such, I was going to propose we move these to the e2epod package (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go).
We could do that, I think, since without `WaitForControlledPodsRunning` and `WaitForControlledPods`, e2epod wouldn;t be importing e2eresource and then e2eresource could import e2epod (more general package calling a specific-task package makes some sense).

Althogh, since `waitForPodsGone` and `waitForPodsInactive` are both private and apparently only used by `DeleteResourceAndWaitForGC` then there is no point in adding the extra complication :smile:

Reply:
@alejandrox1 I think you are right. I prefer to keep the visibility of these methods since they are only used in `DeleteResourceAndWaitForGC` now. And if they become more popular someday, someone could make them public and moved to e2epod .